ray-operator/controllers/ray/batchscheduler/scheduler-plugins/scheduler_plugins.go

```diff
@@ -0,0 +1,154 @@
+	"sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	schedulerinterface "github.com/ray-project/kuberay/ray-operator/controllers/ray/batchscheduler/interface"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+)
+
+const (
+	SchedulerName                 string = "kube-scheduler"
+	KubeSchedulerPodGroupLabelKey string = "scheduling.x-k8s.io/pod-group"
+)
+
+type KubeScheduler struct {
+	cli client.Client
+}
+
+type KubeSchedulerFactory struct{}
+
+func GetPluginName() string {
+	return SchedulerName
+}
+
+func (k *KubeScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (k *KubeScheduler) DoBatchSchedulingOnSubmission(ctx context.Context, rc *rayv1.RayCluster) error {
+	if !k.isGangSchedulingEnabled(rc) {
+		return nil
+	}
+	// we set replica as 1 for the head pod
+	replica := int32(1)
+	for _, workerGroup := range rc.Spec.WorkerGroupSpecs {
+		if workerGroup.Replicas == nil {
+			continue
+		}
+		replica += *workerGroup.Replicas
+	}
+
+	podGroup := &v1alpha1.PodGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: rc.Namespace,
+			Name:      rc.Name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       rc.Name,
+					UID:        rc.UID,
+					APIVersion: rc.APIVersion,
+					Kind:       rc.Kind,
+				},
+			},
+		},
+		Spec: v1alpha1.PodGroupSpec{
+			MinMember:    replica,
+			MinResources: utils.CalculateMinResources(rc),
+		},
+	}
+
+	err := k.cli.Create(ctx, podGroup)
+	if errors.IsAlreadyExists(err) {
+		return nil
+	}
+	return err
+}
+
+// AddMetadataToPod adds essential labels and annotations to the Ray pods
+// the scheduler needs these labels and annotations in order to do the scheduling properly
+func (k *KubeScheduler) AddMetadataToPod(_ context.Context, app *rayv1.RayCluster, _ string, pod *corev1.Pod) {
+	// when gang scheduling is enabled, extra labels need to be added to all pods
+	if k.isGangSchedulingEnabled(app) {
+		pod.Labels[KubeSchedulerPodGroupLabelKey] = app.Name
+	}
+}
+
+func (k *KubeScheduler) isGangSchedulingEnabled(app *rayv1.RayCluster) bool {
+	_, exist := app.Labels[utils.RayClusterGangSchedulingEnabled]
+	return exist
+}
+
+func (kf *KubeSchedulerFactory) New(ctx context.Context, c *rest.Config) (schedulerinterface.BatchScheduler, error) {
+	extClient, err := apiextensionsclient.NewForConfig(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize k8s extension client with error %w", err)
+	}
+
+	podGroupName := "podgroups.scheduling.x-k8s.io"
+	if _, err := extClient.ApiextensionsV1().CustomResourceDefinitions().Get(
+		ctx,
+		podGroupName,
+		metav1.GetOptions{},
+	); err != nil {
+		if _, err := extClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(
+			ctx,
+			podGroupName,
+			metav1.GetOptions{},
+		); err != nil {
+			return nil, fmt.Errorf("podGroup CRD is required to exist in current cluster. error: %w", err)
+		}
+	}
+
+	scheme := runtime.NewScheme()
```

Comment:
can we reuse the controller's informer cache? If this is non-trivial, we don't need to address this in this PR.

TODO: Volcano seems to call K8s API server for each reconciliation.

Reply:
I will create a new issue to track this.